File: KubeVisibility.Dashboard.Api/Models/ClusterInfoModels.cs
namespace KubeVisibility.Dashboard.Api.Models
{
    public class ClusterInfoResponse
    {
        public Dictionary<string, NamespaceInfo> Namespaces { get; set; } = new();
    }

    public class NamespaceInfo
    {
        public bool Success { get; set; }
        public List<ResourceInfo>? Resources { get; set; }
        public string? Error { get; set; }
    }

    public class ResourceInfo
    {
        public string Name { get; set; } = string.Empty;
        public string MetadataName { get; set; } = string.Empty; // Actual Kubernetes resource name for API calls
        public string Type { get; set; } = string.Empty;
        public DateTime? LastUpdated { get; set; }
        public List<ContainerInfo> Containers { get; set; } = new();
        public List<PodInfo> Pods { get; set; } = new();
        
        // Health and Replica Information
        public string HealthStatus { get; set; } = "Unknown"; // Healthy, Degraded, Failed, Unknown
        public int DesiredReplicas { get; set; }
        public int CurrentReplicas { get; set; }
        public int ReadyReplicas { get; set; }
        
        // CronWorkflow Schedule Information
        public string? Schedule { get; set; }
        public DateTime? LastScheduleTime { get; set; }
        
        // Resource Age
        public DateTime? CreationTime { get; set; }
        
        // Recent Events
        public List<EventInfo> RecentEvents { get; set; } = new();
        
        // CronWorkflow Properties (Argo Workflows)
        public List<string>? Schedules { get; set; }
        public int? StartingDeadlineSeconds { get; set; }
        public string? ConcurrencyPolicy { get; set; }
        public int? SuccessfulJobsHistoryLimit { get; set; }
        public int? FailedJobsHistoryLimit { get; set; }
        public bool? Suspend { get; set; }
        
        // Consumer Control (ConfigMap flag)
        public bool? ConsumerEnabled { get; set; }
    }

    public class ContainerInfo
    {
        public string Name { get; set; } = string.Empty;
        public string Image { get; set; } = string.Empty;
        public string Version { get; set; } = string.Empty;
        public Dictionary<string, string> EnvironmentVariables { get; set; } = new();
        
        // Resource Limits
        public string CpuRequest { get; set; } = string.Empty;
        public string CpuLimit { get; set; } = string.Empty;
        public string MemoryRequest { get; set; } = string.Empty;
        public string MemoryLimit { get; set; } = string.Empty;
        public string ImagePullPolicy { get; set; } = string.Empty;
    }

    public class PodInfo
    {
        public string Name { get; set; } = string.Empty;
        public string Status { get; set; } = string.Empty;
        public DateTime? StartTime { get; set; }
        public string? RestartCount { get; set; }
        
        // Pod Health Information
        public bool IsReady { get; set; }
        public string Reason { get; set; } = string.Empty;        // CrashLoopBackOff, ImagePullBackOff, etc.
        public string Message { get; set; } = string.Empty;       // Detailed error message
        public string NodeName { get; set; } = string.Empty;
        
        // Resource Utilization
        public string? CpuUsage { get; set; }           // Current CPU usage (e.g., "100m", "1.5")
        public string? MemoryUsage { get; set; }        // Current memory usage (e.g., "512Mi", "1Gi")
        public string? CpuLimit { get; set; }          // CPU limit from pod spec
        public string? MemoryLimit { get; set; }       // Memory limit from pod spec
        public string? CpuRequest { get; set; }         // CPU request from pod spec
        public string? MemoryRequest { get; set; }      // Memory request from pod spec
        public double? CpuUsagePercent { get; set; }    // CPU usage as percentage of limit
        public double? MemoryUsagePercent { get; set; } // Memory usage as percentage of limit
    }
    
    public class EventInfo
    {
        public string Type { get; set; } = string.Empty;    // Normal, Warning, Error
        public string Reason { get; set; } = string.Empty;
        public string Message { get; set; } = string.Empty;
        public DateTime Timestamp { get; set; }
    }
    
    // Consumer Control Models
    public class ConsumerControlFlagResponse
    {
        public bool Enabled { get; set; }
        public bool Exists { get; set; }
    }
    
    public class ToggleConsumerRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string DeploymentName { get; set; } = string.Empty;
        public bool Enabled { get; set; }
    }
    
    public class ToggleConsumerResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public ResourceInfo? UpdatedResource { get; set; }
    }
    
    // CronWorkflow Management Models
    public class UpdateCronWorkflowRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
        public List<string> Schedules { get; set; } = new();
        public int StartingDeadlineSeconds { get; set; }
        public string ConcurrencyPolicy { get; set; } = string.Empty;
        public int SuccessfulJobsHistoryLimit { get; set; }
        public int FailedJobsHistoryLimit { get; set; }
        public bool Suspend { get; set; }
    }
    
    public class UpdateCronWorkflowResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public ResourceInfo? UpdatedResource { get; set; }
    }

    public class UpdateCronWorkflowLogLevelRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
        public string LogLevel { get; set; } = string.Empty;
    }

    public class UpdateCronWorkflowLogLevelResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public ResourceInfo? UpdatedResource { get; set; }
    }
    
    public class ToggleSuspendRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
        public bool Suspend { get; set; }
    }
    
    public class ToggleSuspendResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public bool? CurrentSuspendState { get; set; }
        public ResourceInfo? UpdatedResource { get; set; }
    }

    // Request models for controllers
    public class RestartPodRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string PodName { get; set; } = string.Empty;
    }

    public class RestartResourceRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
        public string ResourceType { get; set; } = string.Empty;
    }

    public class UpdateLogLevelRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
        public string ResourceType { get; set; } = string.Empty;
        public string LogLevel { get; set; } = string.Empty;
    }

    public class UpdateLogLevelResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
    }
    
    // Argo Workflows Submit Models
    public class SubmitCronWorkflowRequest
    {
        public string NamespaceName { get; set; } = string.Empty;
        public string ResourceName { get; set; } = string.Empty;
    }
    
    public class SubmitCronWorkflowResponse
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public string? WorkflowName { get; set; }
    }
}

File: KubeVisibility.Dashboard.UI/src/app/core/models/cluster-info.models.ts
export interface ClusterInfoResponse {
  namespaces: Record<string, NamespaceInfo>;
}

export interface NamespaceInfo {
  success: boolean;
  resources?: ResourceInfo[];
  error?: string;
}

export interface ResourceInfo {
  name: string;
  metadataName: string;
  type: string;
  lastUpdated?: string;
  containers: ContainerInfo[];
  pods: PodInfo[];
  healthStatus: 'Healthy' | 'Degraded' | 'Failed' | 'Unknown';
  desiredReplicas: number;
  currentReplicas: number;
  readyReplicas: number;
  schedule?: string;
  lastScheduleTime?: string;
  creationTime?: string;
  recentEvents: EventInfo[];
  schedules?: string[];
  startingDeadlineSeconds?: number;
  concurrencyPolicy?: string;
  successfulJobsHistoryLimit?: number;
  failedJobsHistoryLimit?: number;
  suspend?: boolean;
  consumerEnabled?: boolean;
}

export interface ContainerInfo {
  name: string;
  image: string;
  version: string;
  environmentVariables: Record<string, string>;
  cpuRequest: string;
  cpuLimit: string;
  memoryRequest: string;
  memoryLimit: string;
  imagePullPolicy: string;
}

export interface PodInfo {
  name: string;
  status: string;
  startTime?: string;
  restartCount?: string;
  isReady: boolean;
  reason: string;
  message: string;
  nodeName: string;
  // Resource Utilization
  cpuUsage?: string;
  memoryUsage?: string;
  cpuLimit?: string;
  memoryLimit?: string;
  cpuRequest?: string;
  memoryRequest?: string;
  cpuUsagePercent?: number;
  memoryUsagePercent?: number;
}

export interface EventInfo {
  type: string;
  reason: string;
  message: string;
  timestamp: string;
}

export interface ConsumerControlFlagResponse {
  enabled: boolean;
  exists: boolean;
}

export interface ToggleConsumerRequest {
  namespaceName: string;
  deploymentName: string;
  enabled: boolean;
}

export interface ToggleConsumerResponse {
  success: boolean;
  message: string;
  updatedResource?: ResourceInfo;
}

export interface UpdateCronWorkflowRequest {
  namespaceName: string;
  resourceName: string;
  schedules: string[];
  startingDeadlineSeconds: number;
  concurrencyPolicy: string;
  successfulJobsHistoryLimit: number;
  failedJobsHistoryLimit: number;
  suspend: boolean;
}

export interface UpdateCronWorkflowResponse {
  success: boolean;
  message: string;
  updatedResource?: ResourceInfo;
}

export interface UpdateCronWorkflowLogLevelRequest {
  namespaceName: string;
  resourceName: string;
  logLevel: string;
}

export interface UpdateCronWorkflowLogLevelResponse {
  success: boolean;
  message: string;
  updatedResource?: ResourceInfo;
}

export interface ToggleSuspendRequest {
  namespaceName: string;
  resourceName: string;
  suspend: boolean;
}

export interface ToggleSuspendResponse {
  success: boolean;
  message: string;
  currentSuspendState?: boolean;
  updatedResource?: ResourceInfo;
}

export interface SubmitCronWorkflowRequest {
  namespaceName: string;
  resourceName: string;
}

export interface SubmitCronWorkflowResponse {
  success: boolean;
  message: string;
  workflowName?: string;
}

export interface RestartPodRequest {
  namespaceName: string;
  podName: string;
}

export interface RestartResourceRequest {
  namespaceName: string;
  resourceName: string;
  resourceType: string;
}

export interface UpdateLogLevelRequest {
  namespaceName: string;
  resourceName: string;
  resourceType: string;
  logLevel: string;
}

export interface UpdateLogLevelResponse {
  success: boolean;
  message: string;
}

export interface PodLogsResponse {
  logs: string;
}

export type ViewMode = 'scanning' | 'deepdive';
export type HealthFilter = 'healthy' | 'degraded' | 'failed' | 'paused' | 'stopped';
export type HealthFilterSelection = HealthFilter[];

